src/cmap/connection_pool.ts

```diff
@@ -643,6 +650,82 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
     );
   }
 
+  async withConnectionAsync(
+    conn: Connection | undefined,
+    fn: AsyncWithConnectionCallback
+  ): Promise<Connection> {
+    let result;
+    if (conn) {
+      // use the provided connection, and do _not_ check it in after execution
+      try {
+        result = await fn(undefined, conn);
+        return result;
+      } catch (fnErr) {
+        if (fnErr instanceof MongoError && fnErr.code === MONGODB_ERROR_CODES.Reauthenticate) {
+          conn = this.reauthenticateAsync(conn);
+        } else {
+          throw fnErr;
+        }
+      }
+    } else {
+      const { promise, resolve, reject } = promiseWithResolvers<Connection>();
+      this.checkOut((err, conn) => {
+        fn(err as MongoError, conn).then(
+          () => {
+            if (conn) this.checkIn(conn);
+          },
+          fnErr => {
+            if (conn) {
+              this.withReauthenticationAsync(fnErr, conn, fn).then(resolve, reject);
+            } else {
+              reject(fnErr);
+            }
+          }
+        );
+      });
+      return promise;
+    }
+  }
+
+  async withReauthenticationAsync(
+    fnErr: AnyError,
+    conn: Connection,
+    fn: AsyncWithConnectionCallback
+  ): Promise<Connection> {
+    if (fnErr instanceof MongoError && fnErr.code === MONGODB_ERROR_CODES.Reauthenticate) {
+      conn = await this.reauthenticateAsync(conn);
+      return fn(undefined, conn);
+    } else {
+      throw fnErr;
+    }
+  }
+
+  async reauthenticateAsync(connection: Connection): Promise<Connection> {
+    const authContext = connection.authContext;
+    if (!authContext) {
+      throw new MongoRuntimeError('No auth context found on connection.');
+    }
+    const credentials = authContext.credentials;
+    if (!credentials) {
+      throw new MongoMissingCredentialsError(
+        'Connection is missing credentials when asked to reauthenticate'
+      );
+    }
+
+    const resolvedCredentials = credentials.resolveAuthMechanism(connection.hello);
+    const provider = AUTH_PROVIDERS.get(resolvedCredentials.mechanism);
+
+    if (!provider) {
+      throw new MongoMissingCredentialsError(
+        `Reauthenticate failed due to no auth provider for ${credentials.mechanism}`
+      );
+    }
+
+    await provider.reauth(authContext);
+
+    return connection;
```

Comment:
NOTE: This function is actually used in Server.command to implement reauth logic